### ISSUE TITLE
Aplicação: Weather Application

### ISSUE THREAD

**Issue by alexvieirasj** [roger-melo-treinamentos/curso-de-js-roger-melo#1917]
### As cores da sua versão da aplicação são diferentes das cores da aplicação mostrada na aula?

Não

### A aplicação contém funcionalidades que não foram mostradas nas aulas?

Não

### A aplicação contém features da linguagem que não foram mostradas nas aulas?

Não

### Link do repositório da aplicação

https://github.com/alexvieirasj/weather-application

### Maiores dificuldades durante a implementação

São 2 itens: primeiro foi entender o fluxo de como a API funciona, e testar isoladamento a parte do JS e depois ir implementar no HTML (conforme mostrado nas aulas). O segundo é refatorar a aplicação sem quebrar é o mais complicado, mas isso requear bastante prática e exercícios.

### Menores dificuldades durante a implementação

Utilizar o Javascript, HTML e CSS.

**Comment by alexvieirasj** [comment]
Certo, obrigado pela orientação, salvei o e-mail. Fiz a alteração necessária e fiz o commit para o meu repositório. Pode prosseguir com as correções. Qualquer coisa fico a disposição.

**Comment by Roger-Melo** [comment]
Olá @alexvieirasj!

Parabéns pelo esforço em construir a terceira aplicação do treinamento =)

> ### São 2 itens: primeiro foi entender o fluxo de como a API funciona, e testar isoladamento a parte do JS e depois ir implementar no HTML (conforme mostrado nas aulas). 

Geralmente, APIs mais complexas ou com volume maior de dados maiores vão exigir mais requests. Essa API de previsão do tempo não é tão complexa mas dá uma boa introdução à requests sequenciais =)

> ### O segundo é refatorar a aplicação sem quebrar é o mais complicado, mas isso requear bastante prática e exercícios.

Exatamente. 

O mais importante no primeiro momento vc já fez, que é fazer funcionar. Suas habilidades de refatoração evoluem com a prática e o tempo, ok?

Vou deixar abaixo alguns pontos, segundo os critérios que usei para analisar.

---

## Boas decisões

Você usou bem o `autofocus` e o `reset`, evitando que o usuário tenha que tocar ou clicar no input.

A forma como vc executou ternário (sem efeito colateral) demonstra que vc entendeu como usá-lo.

Nas análises passadas vc já havia demonstrado um bom entendimento da lógica. E dessa vez não foi diferente =)

## Organização de código

### Comentários

Não é uma regra, mas evite comentários.

Comentários podem mentir. Se o código for modificado e o comentário correspondente não for atualizado, o comentário perde o sentido. Isso sem mencionar o retrabalho de atualizar comentários.

### Refatoração

Se vc quiser, pode nomear o callback do `addEventListener` `cityForm` e reorganizar algumas linhas do código.

Pra fazer isso, vc pode mover o código do callback para a função `showCityWeather` e apenas referenciá-la. Ficaria mais ou menos assim:

```javascript
const showCityWeather = async event => {
  event.preventDefault()
  
  const [{ Key, LocalizedName }] = await getCityData(event.target.city.value)
  const [{ WeatherText, Temperature, IsDayTime, WeatherIcon }] = await getCityWeather(Key)
  const timeIcon = `<img src="./src/icons/${WeatherIcon}.svg" />`

  timeImage.src = IsDayTime ? './src/day.svg' : './src/night.svg'
  
  showFields(timeIcon, LocalizedName, WeatherText, Temperature)
  showCity()
  cityForm.reset()
}

cityForm.addEventListener('submit', showCityWeather)
```

Para deixar a `showCity` mais específica, eu a nomearia como `showCard` ou `showCityCard`.

Observe também que `showFields` pode receber um objeto ao invés de vários argumentos/parâmetros.

Você pode usar um objeto como argumento de uma função quando quiser:

- Diminuir o número de argumentos/parâmetros que uma função recebe;
- Deixar mais claro o que os valores inseridos como argumentos significam.

Funções com vários argumentos são mais complexas de testar e manter, pois exigem que casos diferentes sejam testados com cada argumento individualmente.

Como o JS permite que objetos sejam criados à qualquer momento, sem a necessidade de usar classes, um objeto pode substituir vários argumentos espalhados na invocação de uma função.

Outra vantagem é que objetos não precisam obedecer uma ordem de parâmetros. Na função abaixo, é obrigatório que ao invocá-la, o ícone seja o 1º argumento, o nome da cidade o segundo, etc. Se essa ordem não for obedecida, o código quebra.

```javascript
const showFields = (timeIcon, LocalizedName, WeatherText, Temperature) => {
  // ...
}
```

Pegando um outro exemplo, se vc observar a invocação abaixo, não fica tão explícito o que os valores inseridos como argumentos significam:

```javascript
manipulateClasses(todosToHide, 'hidden', 'd-flex')
```

Porém, se um objeto for inserido como argumento, fica mais fácil entender por que 'hidden' e 'd-flex' (por exemplo) estão sendo usadas ali:

```javascript
manipulateClasses({ todos: todosToHide, classToAdd: 'hidden', classToRemove: 'd-flex' })
```

Ou seja, fica fácil saber que `hidden` é a classe a ser adicionada e `d-flex` é a classe a ser removida. Mas claro, isso só funciona se os nomes das propriedades deixarem isso explícito.

Voltando para o exemplo da `showFields`, mesmo que os argumentos que ela recebe estejam legíveis, acho que vale a refatoração, devido às vantagens que mencionei acima.

Ficaria assim:

```javascript
const showFields = fields => {
  // ...
};

// ...

showFields({ timeIcon, LocalizedName, WeatherText, Temperature })
```

E claro, se vc quiser, pode fazer o destructuring do objeto que está sendo recebido como parâmetro:

```javascript
const showFields = ({ timeIcon, LocalizedName, WeatherText, Temperature }) => {
  // ...
}
```

---

Nas aulas seguintes, você verá outras formas de refatorar o código dessa aplicação.

Mais uma vez, parabéns pelo esforço =)

O que eu falei fez sentido?

**Comment by alexvieirasj** [comment]
Com certeza, todos os apontamentos são de grande valia, com certeza vou refatorar o código para seguir um padrão de boas práticas e também para assimilar e evoluir como programador. Mais uma vez obrigado pelas dicas e orientações.
